File: NanoAnalysis/python/modules/MuonScaRe.py
'''
  Credit: J. Von Den Driesch
  Taken from: https://gitlab.cern.ch/cms-muonPOG/muonscarekit
'''

import numpy as np
from scipy.special import erfinv, erf
from random import random
import awkward as ak

class CrystallBall:

    def __init__(self, m=0, s=1, a=10, n=10):
        self.pi = 3.14159
        self.sqrtPiOver2 = np.sqrt(self.pi/2.0)
        self.sqrt2 = np.sqrt(2.0)
        self.m = ak.Array(m)
        self.s = ak.Array(s)
        self.a = ak.Array(a)
        self.n = ak.Array(n)
        self.fa = abs(self.a)
        self.ex = np.exp(-self.fa * self.fa/2)
        self.A  = (self.n/self.fa)**self.n * self.ex
        self.C1 = self.n/self.fa/(self.n-1) * self.ex
        self.D1 = 2 * self.sqrtPiOver2 * erf(self.fa/self.sqrt2)

        self.B = self.n/self.fa - self.fa
        self.C = (self.D1 + 2 * self.C1)/self.C1
        self.D = (self.D1 + 2 * self.C1)/2

        self.N = 1.0/self.s/(self.D1 + 2 * self.C1)
        self.k = 1.0/(self.n - 1)

        self.NA = self.N * self.A
        self.Ns = self.N * self.s
        self.NC = self.Ns * self.C1
        self.F = 1 - self.fa * self.fa/self.n
        self.G = self.s * self.n/self.fa
        self.cdfMa = self.cdf(self.m - self.a * self.s)
        self.cdfPa = self.cdf(self.m + self.a * self.s)

    def cdf(self, x):
        x = ak.Array(x)
        d = (x - self.m)/self.s
        result = ak.full_like(d, 1.0)
        
        # define different conditions
        c1a = (d<-self.a) & (self.F - self.s * d / self.G > 0)
        c1b = (d<-self.a) & (self.F - self.s * d / self.G <= 0)
        c2a = (d>self.a) & (self.F + self.s * d / self.G > 0)
        c2b = (d>self.a) & (self.F + self.s * d / self.G <= 0)

        c3 = ~c1a & ~c1b & ~c2a & ~c2b
        # For d < -a
        result = ak.where(c1a, 
                        self.NC / np.power(self.F - self.s * d / self.G, self.n - 1), 
                        result)
        result = ak.where(c1b, 
                        self.NC, 
                        result)

        # For d > a
        result = ak.where(c2a, 
                        self.NC * (self.C - np.power(self.F + self.s * d / self.G, 1 - self.n)), 
                        result)
        result = ak.where(c2b, 
                        self.NC * self.C, 
                        result)

        # For -a <= d <= a
        result = ak.where(c3, 
                        self.Ns * (self.D - self.sqrtPiOver2 * erf(-d / self.sqrt2)), 
                        result)

        return result



    def invcdf(self, u):
        u = ak.Array(u)
        result = ak.zeros_like(u)

        c1a = (u < self.cdfMa) & (self.NC/u > 0)
        c1b = (u < self.cdfMa) & (self.NC/u <= 0)
        c2a = (u > self.cdfPa) &  (self.C - u/self.NC > 0)
        c2b = (u > self.cdfPa) &  (self.C - u/self.NC <= 0)
        c3 = ~c1a & ~c1b & ~c2a & ~c2b

        # For u < cdfMa
        result = ak.where(c1a, 
                        self.m + self.G * (self.F - (self.NC / u) ** self.k), 
                        result)
        result = ak.where(c1b, 
                        self.m + self.G * self.F, 
                        result)

        # For u > cdfPa
        result = ak.where(c2a, 
                        self.m - self.G * (self.F - (self.C - u / self.NC) ** (-self.k)), 
                        result)
        result = ak.where(c2b, 
                        self.m - self.G * self.F, 
                        result)

        # For cdfMa <= u <= cdfPa
        result = ak.where(c3, 
                        self.m - self.sqrt2 * self.s * erfinv((self.D - u / self.Ns) / self.sqrtPiOver2), 
                        result)

        return result


def get_rndm(eta, nL, cset, nested=False):
    # obtain parameters from correctionlib
    if nested:
        eta_f, nL_f, nmuons = ak.flatten(eta), ak.flatten(nL), ak.num(nL)
    else:
        eta_f, nL_f, nmuons = eta, nL, np.ones_like(eta)

    mean_f = cset.get("cb_params").evaluate(abs(eta_f), nL_f, 0)
    sigma_f = cset.get("cb_params").evaluate(abs(eta_f), nL_f, 1)
    n_f = cset.get("cb_params").evaluate(abs(eta_f), nL_f, 2)
    alpha_f = cset.get("cb_params").evaluate(abs(eta_f), nL_f, 3)

    # get random number following the CB
    # print(nmuons)
    rndm_f = [random() for i in nmuons for j in range(int(i))]

    cb_f = CrystallBall(mean_f, sigma_f, alpha_f, n_f)

    result_f = cb_f.invcdf(rndm_f)

    if nested:
        result = ak.unflatten(result_f, nmuons)
    else:
        result = result_f

    return result


def get_std(pt, eta, nL, cset, nested=False):
    if nested:
        eta_f, nL_f, pt_f, nmuons = ak.flatten(eta), ak.flatten(nL), ak.flatten(pt), ak.num(nL)
    else:
        eta_f, nL_f, pt_f, nmuons = eta, nL, pt, 1

    # obtain parameters from correctionlib    
    param0_f = cset.get("poly_params").evaluate(abs(eta_f), nL_f, 0)
    param1_f = cset.get("poly_params").evaluate(abs(eta_f), nL_f, 1)
    param2_f = cset.get("poly_params").evaluate(abs(eta_f), nL_f, 2)

    # calculate value and return max(0, val)
    sigma_f = param0_f + param1_f * pt_f + param2_f * pt_f*pt_f
    sigma_corrected_f = np.where(sigma_f < 0, 0, sigma_f)

    if nested:
        result = ak.unflatten(sigma_corrected_f, nmuons)
    else:
        result = sigma_corrected_f

    return result


def get_k(eta, var, cset, nested=False):
    if nested:
        eta_f, nmuons = ak.flatten(eta), ak.num(eta)
    else:
        eta_f = eta

    # obtain parameters from correctionlib
    k_data_f = cset.get("k_data").evaluate(abs(eta_f), var)
    k_mc_f = cset.get("k_mc").evaluate(abs(eta_f), var)

    # calculate residual smearing factor 
    # return 0 if smearing in MC already larger than in data
    k_f = np.zeros_like(k_data_f)
    condition = k_mc_f<k_data_f

    k_f[condition] = (k_data_f[condition]**2 - k_mc_f[condition]**2)**.5

    if nested:
        result = ak.unflatten(k_f, nmuons)
    else:
        result = k_f

    return result


def pt_resol(pt, eta, nL, var, cset, nested=False):
    """"
    Function for the calculation of the resolution correction
    Input: 
    pt - muon transverse momentum
    eta - muon pseudorapidity
    nL - muon number of tracker layers
    var - variation (standard is "nom")
    cset - correctionlib object

    This function should only be applied to reco muons in MC!
    """
    rndm = get_rndm(eta, nL, cset, nested)
    std = get_std(pt, eta, nL, cset, nested)
    k = get_k(eta, var, cset, nested)

    pt_corr = pt * (1 + k * std * rndm)
    pt_corr_corr = np.where(np.isnan(pt_corr), pt, pt_corr)

    return pt_corr_corr


def pt_scale(is_data, pt, eta, phi, charge, var, cset, nested=False):
    """
    Function for the calculation of the scale correction
    Input:
    is_data - flag that is True if dealing with data and False if MC
    pt - muon transverse momentum
    eta - muon pseudorapidity
    phi - muon angle
    charge - muon charge
    var - variation (standard is "nom")
    cset - correctionlib object
    
    This function should be applied to reco muons in data and MC
    """
    if is_data:
        dtmc = "data"
    else:
        dtmc = "mc"

    if nested:
        eta_f, phi_f, nmuons = ak.flatten(eta), ak.flatten(phi), ak.num(eta)
    else:
        eta_f, phi_f, nmuons = eta, phi, 1
    
    a_f = cset.get("a_"+dtmc).evaluate(eta_f, phi_f, var)
    m_f = cset.get("m_"+dtmc).evaluate(eta_f, phi_f, var)

    if nested:
        a, m = ak.unflatten(a_f, nmuons), ak.unflatten(m_f, nmuons)
    else: 
        a, m = a_f, m_f

    return 1. / (m/pt + charge * a)

File: AnalysisStep/src/MuonScaRe.cc
/*  
 * Run3 Muon correction module, imported from:
 * https://gitlab.cern.ch/cms-muonPOG/muonscarekit/-/blob/82cc2a4ec97c93ee4e791398f4bc5fbe24859bb3/scripts/MuonScaRe.cc
 * with the following modifications:
 * -encapsulated in a class so that the correctionSet and RNG are kept 
 *  internally
 * -add function to optionally set the RNG seed, so that the user can implement 
 *  deterministic pseudo-random smearing. If desired, a proper seed should 
 *  be passed once for each muon before calling pt_resol the first time for 
 *  that mu.
 *  Note: setting the seed is left to the user since in the current 
 *  implementation get_rndm, which normally getsis called 3 times for each mu 
 *  (for var = norm, syst, stat), does not have access to enough sources of 
 *  entropy to create a decent seed internally.
 * -avoid trivial string manipulations in pt_scale
 */

#include <boost/math/special_functions/erf.hpp>
#include <string>
#include <ZZAnalysis/AnalysisStep/interface/MuonScaRe.h>

using namespace std;

MuonScaRe::MuonScaRe(string json) : cset(correction::CorrectionSet::from_file(json)){}
 

struct CrystalBall{
    double pi=3.14159;
    double sqrtPiOver2=sqrt(pi/2.0);
    double sqrt2=sqrt(2.0);
    double m;
    double s;
    double a;
    double n;
    double B;
    double C;
    double D;
    double N;
    double NA;
    double Ns;
    double NC;
    double F;
    double G;
    double k;
    double cdfMa;
    double cdfPa;
CrystalBall():m(0),s(1),a(10),n(10){
    init();
}
CrystalBall(double mean, double sigma, double alpha, double n)
    :m(mean),s(sigma),a(alpha),n(n){
    init();
}
void init(){
    double fa = fabs(a);
    double ex = exp(-fa*fa/2);
    double A  = pow(n/fa, n) * ex;
    double C1 = n/fa/(n-1) * ex; 
    double D1 = 2 * sqrtPiOver2 * erf(fa/sqrt2);
    B = n/fa-fa;
    C = (D1+2*C1)/C1;   
    D = (D1+2*C1)/2;   
    N = 1.0/s/(D1+2*C1); 
    k = 1.0/(n-1);  
    NA = N*A;       
    Ns = N*s;       
    NC = Ns*C1;     
    F = 1-fa*fa/n; 
    G = s*n/fa;    
    cdfMa = cdf(m-a*s);
    cdfPa = cdf(m+a*s);
}
double pdf(double x) const{ 
    double d=(x-m)/s;
    if(d<-a) return NA*pow(B-d, -n);
    if(d>a) return NA*pow(B+d, -n);
    return N*exp(-d*d/2);
}
double pdf(double x, double ks, double dm) const{ 
    double d=(x-m-dm)/(s*ks);
    if(d<-a) return NA/ks*pow(B-d, -n);
    if(d>a) return NA/ks*pow(B+d, -n);
    return N/ks*exp(-d*d/2);

}
double cdf(double x) const{
    double d = (x-m)/s;
    if(d<-a) return NC / pow(F-s*d/G, n-1);
    if(d>a) return NC * (C - pow(F+s*d/G, 1-n) );
    return Ns * (D - sqrtPiOver2 * erf(-d/sqrt2));
}
double invcdf(double u) const{
    if(u<cdfMa) return m + G*(F - pow(NC/u, k));
    if(u>cdfPa) return m - G*(F - pow(C-u/NC, -k) );
    return m - sqrt2 * s * boost::math::erf_inv((D - u/Ns )/sqrtPiOver2);
}
};

double MuonScaRe::get_rndm(double eta, float nL) {

    // obtain parameters from correctionlib
    double mean = cset->at("cb_params")->evaluate({abs(eta), nL, 0});
    double sigma = cset->at("cb_params")->evaluate({abs(eta), nL, 1});
    double n = cset->at("cb_params")->evaluate({abs(eta), nL, 2});
    double alpha = cset->at("cb_params")->evaluate({abs(eta), nL, 3});
    
    // instantiate CB and get random number following the CB
    CrystalBall cb(mean, sigma, alpha, n);
    return cb.invcdf(rng.Rndm());
}


double MuonScaRe::get_std(double pt, double eta, float nL) {

    // obtain paramters from correctionlib
    double param_0 = cset->at("poly_params")->evaluate({abs(eta), nL, 0});
    double param_1 = cset->at("poly_params")->evaluate({abs(eta), nL, 1});
    double param_2 = cset->at("poly_params")->evaluate({abs(eta), nL, 2});

    // calculate value and return max(0, val)
    double sigma = param_0 + param_1 * pt + param_2 * pt*pt;
    if (sigma < 0) sigma = 0;
    return sigma; 
}


double MuonScaRe::get_k(double eta, string var) {

    // obtain parameters from correctionlib
    double k_data = cset->at("k_data")->evaluate({abs(eta), var});
    double k_mc = cset->at("k_mc")->evaluate({abs(eta), var});

    // calculate residual smearing factor
    // return 0 if smearing in MC already larger than in data
    double k = 0;
    if (k_mc < k_data) k = sqrt(k_data*k_data - k_mc*k_mc);
    return k;
}


double MuonScaRe::pt_resol(double pt, double eta, float nL, string var) {

    // load correction values
    double rndm = (double) get_rndm(eta, nL);
    double std = (double) get_std(pt, eta, nL);
    double k = (double) get_k(eta, var);

    // calculate corrected value and return original value if a parameter is nan
    double ptc = pt * ( 1 + k * std * rndm);
    if (isnan(ptc)) ptc = pt;
    return ptc;
}


double MuonScaRe::pt_scale(bool is_data, double pt, double eta, double phi, int charge, string var) {
  
    double a = cset->at(is_data?"a_data":"a_mc")->evaluate({eta, phi, var});
    double m = cset->at(is_data?"m_data":"m_mc")->evaluate({eta, phi, var});
    return 1. / (m/pt + charge * a);
}
